File: typescript/16-proyecto/fase4_hitl.ts
// Fase 4: añade checkpoint HITL para hallazgos críticos.
// En producción: checkpoint → webhook/Slack; aquí: CLI interactiva.

// Cómo ejecutar: make ts SCRIPT=typescript/16-proyecto/fase4_hitl.ts

import * as crypto from "crypto";
import * as fs from "fs";
import * as readline from "readline";

import { agenteRevisionConMemoria } from "./fase3_memoria";
import { ejecutarHerramienta, Hallazgo, RevisionOutput } from "./fase2_herramientas";

interface RevisionConMeta extends RevisionOutput {
  cached?: boolean;
  hitl_descarte?: string;
}

interface ResultadoAprobacion {
  aprobado: boolean;
  revision: RevisionConMeta;
}

function necesitaAprobacion(revision: RevisionConMeta): boolean {
  return revision.hallazgos.some((h) => h.severidad === "critical");
}

async function pregunta(rl: readline.Interface, texto: string): Promise<string> {
  return new Promise((resolve) => rl.question(texto, resolve));
}

async function solicitarAprobacionCli(
  revision: RevisionConMeta
): Promise<ResultadoAprobacion> {
  const criticos = revision.hallazgos.filter((h) => h.severidad === "critical");

  console.log("\n=== REVISIÓN REQUIERE APROBACIÓN ===");
  console.log(`Se encontraron ${criticos.length} hallazgo(s) crítico(s):\n`);

  criticos.forEach((h, i) => {
    console.log(`${i + 1}. Línea ${h.linea ?? "?"}: ${h.descripcion}`);
    console.log(`   Sugerencia: ${h.sugerencia}\n`);
  });

  console.log("Opciones:");
  console.log("  [a] Aprobar y emitir informe completo");
  console.log("  [m] Modificar un hallazgo antes de emitir");
  console.log("  [d] Descartar hallazgos críticos con justificación");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  try {
    const opcion = (await pregunta(rl, "\nElige [a/m/d]: ")).trim().toLowerCase();

    if (opcion === "a") {
      return { aprobado: true, revision };
    }

    if (opcion === "m") {
      const idxStr = await pregunta(
        rl,
        `Número de hallazgo a modificar (1-${criticos.length}): `
      );
      const idx = parseInt(idxStr, 10) - 1;
      const nuevaDesc = await pregunta(rl, "Nueva descripción: ");
      criticos[idx].descripcion = nuevaDesc;
      return { aprobado: true, revision };
    }

    if (opcion === "d") {
      const justificacion = await pregunta(rl, "Justificación para descartar: ");
      revision.hallazgos = revision.hallazgos.filter(
        (h) => h.severidad !== "critical"
      );
      revision.hitl_descarte = justificacion;
      return { aprobado: true, revision };
    }

    return { aprobado: false, revision };
  } finally {
    rl.close();
  }
}

export async function pipelineRevisionCompleto(
  codigo: string,
  ruta: string,
  proyectoDir: string
): Promise<RevisionConMeta | { estado: string; revision: null }> {
  let revision: RevisionConMeta = await agenteRevisionConMemoria(
    codigo,
    ruta,
    proyectoDir
  );

  if (necesitaAprobacion(revision)) {
    const resultado = await solicitarAprobacionCli(revision);
    if (!resultado.aprobado) {
      return { estado: "rechazado", revision: null };
    }
    revision = resultado.revision;
  }

  const nombreInforme = `revision_${crypto
    .createHash("md5")
    .update(ruta)
    .digest("hex")
    .slice(0, 8)}.json`;

  ejecutarHerramienta(
    "write_report",
    {
      content: JSON.stringify(revision, null, 2),
      filename: nombreInforme,
    },
    proyectoDir
  );

  return revision;
}

async function main() {
  const args = process.argv.slice(2);
  let codigo: string;
  let ruta: string;
  let proyectoDir: string;

  if (args.length > 0) {
    codigo = require("fs").readFileSync(args[0], "utf-8");
    ruta = args[0];
    proyectoDir = args[1] || process.cwd();
  } else {
    codigo = `
import subprocess

def ejecutar_comando(cmd_usuario):
    # CRÍTICO: inyección de comandos — nunca hacer esto
    resultado = subprocess.run(cmd_usuario, shell=True, capture_output=True, text=True)
    return resultado.stdout
`;
    ruta = "test.py";
    proyectoDir = process.cwd();
  }

  const resultado = await pipelineRevisionCompleto(codigo, ruta, proyectoDir);
  console.log(JSON.stringify(resultado, null, 2));
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: python/16-proyecto/fase4_hitl.py
# Fase 4: añade checkpoint HITL para hallazgos críticos.
# En producción: checkpoint → webhook/Slack; aquí: CLI interactiva.
#
# Cómo ejecutar:
#   make py SCRIPT=python/16-proyecto/fase4_hitl.py
#
# Qué esperar:
#   El agente revisa codigo y pide aprobacion humana para hallazgos criticos.
#   El humano puede aprobar, rechazar o modificar cada accion critica via CLI.
#
# Variables de entorno:
#   MODEL — modelo a usar (default: claude-sonnet-4-6)

import json
import hashlib
import os

from fase3_memoria import agente_revision_con_memoria
from fase2_herramientas import ejecutar_herramienta


def necesita_aprobacion(revision: dict) -> bool:
    hallazgos_criticos = [
        h for h in revision["hallazgos"]
        if h["severidad"] == "critical"
    ]
    return len(hallazgos_criticos) > 0


def solicitar_aprobacion_cli(revision: dict) -> dict:
    """Checkpoint HITL en terminal. En producción: webhook, Slack, UI."""
    criticos = [h for h in revision["hallazgos"] if h["severidad"] == "critical"]

    print("\n=== REVISIÓN REQUIERE APROBACIÓN ===")
    print(f"Se encontraron {len(criticos)} hallazgo(s) crítico(s):\n")

    for i, h in enumerate(criticos, 1):
        print(f"{i}. Línea {h.get('linea', '?')}: {h['descripcion']}")
        print(f"   Sugerencia: {h['sugerencia']}\n")

    print("Opciones:")
    print("  [a] Aprobar y emitir informe completo")
    print("  [m] Modificar un hallazgo antes de emitir")
    print("  [d] Descartar hallazgos críticos con justificación")

    opcion = input("\nElige [a/m/d]: ").strip().lower()

    if opcion == "a":
        return {"aprobado": True, "revision": revision}

    elif opcion == "m":
        idx = int(input(f"Número de hallazgo a modificar (1-{len(criticos)}): ")) - 1
        nueva_desc = input("Nueva descripción: ")
        criticos[idx]["descripcion"] = nueva_desc
        return {"aprobado": True, "revision": revision}

    elif opcion == "d":
        justificacion = input("Justificación para descartar: ")
        revision["hallazgos"] = [h for h in revision["hallazgos"] if h["severidad"] != "critical"]
        revision["hitl_descarte"] = justificacion
        return {"aprobado": True, "revision": revision}

    return {"aprobado": False, "revision": revision}


def pipeline_revision_completo(codigo: str, ruta: str, proyecto_dir: str) -> dict:
    """Pipeline completo: loop → memoria → HITL → informe."""
    revision = agente_revision_con_memoria(codigo, ruta, proyecto_dir)

    if necesita_aprobacion(revision):
        resultado = solicitar_aprobacion_cli(revision)
        if not resultado["aprobado"]:
            return {"estado": "rechazado", "revision": None}
        revision = resultado["revision"]

    nombre_informe = f"revision_{hashlib.md5(ruta.encode()).hexdigest()[:8]}.json"
    ejecutar_herramienta("write_report", {
        "content": json.dumps(revision, indent=2, ensure_ascii=False),
        "filename": nombre_informe
    }, proyecto_dir)

    return revision


if __name__ == "__main__":
    import sys
    codigo = open(sys.argv[1]).read() if len(sys.argv) > 1 else """
import subprocess

def ejecutar_comando(cmd_usuario):
    # CRÍTICO: inyección de comandos — nunca hacer esto
    resultado = subprocess.run(cmd_usuario, shell=True, capture_output=True, text=True)
    return resultado.stdout
"""
    ruta = sys.argv[1] if len(sys.argv) > 1 else "test.py"
    proyecto = sys.argv[2] if len(sys.argv) > 2 else os.getcwd()

    resultado = pipeline_revision_completo(codigo, ruta, proyecto)
    print(json.dumps(resultado, indent=2, ensure_ascii=False))
